app/models/katello/glue/candlepin/pool.rb

```diff
@@ -27,28 +27,56 @@ def self.included(base)
     end
 
     module ClassMethods
-      def candlepin_data(cp_id)
+      def candlepin_data(cp_id, rescue_gone = false)
         Katello::Resources::Candlepin::Pool.find(cp_id)
+      rescue Katello::Errors::CandlepinPoolGone => e
+        raise e unless rescue_gone
+
+        if (pool_id = ::Katello::Pool.find_by_cp_id(cp_id)&.id)
+          Katello::EventQueue.push_event(::Katello::Events::DeletePool::EVENT_TYPE, pool_id)
+          Rails.logger.warn("Sending pool delete event for missing candlepin pool cp_id=#{cp_id}")
+        end
+        {}
       end
 
       def get_for_owner(organization)
         Katello::Resources::Candlepin::Pool.get_for_owner(organization, true)
       end
 
       def import_pool(cp_pool_id, index_hosts = true)
-        json = candlepin_data(cp_pool_id)
-        ::Katello::Util::Support.active_record_retry do
-          pool = Katello::Pool.where(:cp_id => cp_pool_id, :organization => Organization.find_by(:label => json['owner']['key'])).first_or_create
-          pool.backend_data = json
-          pool.import_data(index_hosts)
+        Katello::Logging.time("import candlepin pool", data: { cp_id: cp_pool_id }) do
+          json = candlepin_data(cp_pool_id)
+
+          org = Organization.find_by(label: json['owner']['key'])
+          fail("Organization with label #{json['owner']['key']} wasn't found while importing Candlepin pool") unless org
```

Comment:
I wonder if their db has orgs with labels that don't match the name? And if we should do any testing around that?

Reply:
Thank you for thinking at that level here. That should never be the case because org label cannot change - and the candlepin Owner should always match because of that. Checked their db - everything matches as I'd expect!